File: web_app/lib/models/assets.ts
import { z } from "zod";
import { ModalityTypeEnum } from "./shared";

export const AnnotateFieldModelSchema = z.object({
  name: z.string(),
  modality: ModalityTypeEnum,
  description: z.string(),
});

export type AnnotateFieldModel = z.infer<typeof AnnotateFieldModelSchema>;

export const AssetModelSchema = z.object({
  id: z.string(),
  dataset_id: z.string(),
  dataset_batch_key: z.string(),
  modality: ModalityTypeEnum,
  name: z.string(),
  description: z.string(),
  created_at: z.string(),
  updated_at: z.string(),
  annotate_fields: z.array(AnnotateFieldModelSchema),
  tags: z.array(z.string()).default([]),
  total_publishes: z.number().default(0),
});

export type AssetModel = z.infer<typeof AssetModelSchema>;

export const AssetCreateModelSchema = z.object({
  dataset_id: z.string().min(1, "Dataset is required"),
  name: z.string().min(1, "Name is required"),
  description: z.string().min(1, "Description is required"),
  annotate_fields: z.array(AnnotateFieldModelSchema).min(1, "At least one field is required"),
  tags: z.array(z.string()).default([]),
});

export type AssetCreateModel = z.infer<typeof AssetCreateModelSchema>;

File: infra/src/scripts/internal/src/internal/database/models/assets.py
from internal.utilities.parser import make_id
from pydantic import BaseModel, computed_field, Field
from datetime import datetime

from .shared import ModalityTypeEnum


class AnnotateFieldModel(BaseModel):
    name: str
    modality: ModalityTypeEnum
    description: str


class AssetModel(BaseModel):
    id: str
    dataset_id: str
    dataset_batch_key: str
    modality: ModalityTypeEnum
    name: str
    description: str
    created_at: datetime
    updated_at: datetime
    annotate_fields: list[AnnotateFieldModel]
    tags: list[str] = []
    total_publishes: int = 0


class AssetCreateModel(BaseModel):
    dataset_id: str
    name: str
    description: str
    annotate_fields: list[AnnotateFieldModel]
    tags: list[str] = []

    @computed_field
    @property
    def id(self) -> str:
        return make_id(self.name)

    @computed_field
    @property
    def created_at(self) -> datetime:
        return datetime.now()

    @computed_field
    @property
    def updated_at(self) -> datetime:
        return datetime.now()


class AssetPublishCreateModel(BaseModel):
    asset_id: str
    publisher_id: str

    @computed_field
    @property
    def created_at(self) -> datetime:
        return datetime.now()

    @computed_field
    @property
    def updated_at(self) -> datetime:
        return datetime.now()


class AssetPublishModel(BaseModel):
    asset_id: str
    publisher_id: str
    is_verified: bool = False
    is_published: bool = False
    publish_key: str
    created_at: datetime
    updated_at: datetime


class AssetPublishBodyModel(BaseModel):
    asset_id: str
